torchtnt/framework/callbacks/torchsnapshot_saver.py

```diff
@@ -243,6 +243,15 @@ def on_train_end(self, state: State, unit: TTrainUnit) -> None:
         snapshot_path = _get_snapshot_save_path(
             self._dirpath, epoch, num_steps_completed
         )
+
+        fs = get_filesystem(snapshot_path)
+        if fs.exists(snapshot_path):
```

Comment:
does checking for solely the snapshot metadata work instead? It would save an extra fsspec call for remote file systems.

Also, this check is running once per rank, which can lead to unwanted network usage for large distributed jobs. it might be better to check from rank 0 and broadcast the result to all ranks.

Reply:
> does checking for solely the snapshot metadata work instead?
it doesn't since the snapshot_path folder can be created by another rank beforehand, causing the other ranks to think the snapshot is already there when in reality it has just started being written to

> Also, this check is running once per rank, which can lead to unwanted network usage for large distributed jobs. it might be better to check from rank 0 and broadcast the result to all ranks.
good point, since this is only called once it may be worth making this check only on rank 0